### PR TITLE
fixes #103: Wrong FS error while import data in RedShift

### LINKED ISSUE #103: Wrong FS error while import data in RedShift
Hi everybody!
I have a simple spark job that gets the data from Neo4j and tries to ingest into RedShift:

```scala
spark.sparkContext.hadoopConfiguration.set("fs.s3a.access.key", awss3accessid.get)
spark.sparkContext.hadoopConfiguration.set("fs.s3a.secret.key", awss3accessky.get)

val neo4jDF = ...

neo4jDF.write
      .format("io.github.spark_redshift_community.spark.redshift")
      .option("url", awsredshifturl.get)
      .option("dbtable", awsredshifttable.get)
      .option("tempdir", awss3tmpdir.get)
      .option("forward_spark_s3_credentials", "true")
      .save()
```

But I get the following error:

```bash
java.lang.IllegalArgumentException: Wrong FS s3://<my_bucket>//<dir>/manifest.json -expected s3a://<my_bucket>

	at org.apache.hadoop.fs.s3native.S3xLoginHelper.checkPath(S3xLoginHelper.java:224)
	at org.apache.hadoop.fs.s3a.S3AFileSystem.checkPath(S3AFileSystem.java:1155)
	at org.apache.hadoop.fs.FileSystem.makeQualified(FileSystem.java:666)
	at org.apache.hadoop.fs.s3a.S3AFileSystem.makeQualified(S3AFileSystem.java:1117)
	at org.apache.hadoop.fs.s3a.S3AFileSystem.qualify(S3AFileSystem.java:1143)
	at org.apache.hadoop.fs.s3a.S3AFileSystem.create(S3AFileSystem.java:1318)
	at org.apache.hadoop.fs.FileSystem.create(FileSystem.java:1195)
	at org.apache.hadoop.fs.FileSystem.create(FileSystem.java:1175)
	at org.apache.hadoop.fs.FileSystem.create(FileSystem.java:1064)
	at org.apache.hadoop.fs.FileSystem.create(FileSystem.java:1052)
	at io.github.spark_redshift_community.spark.redshift.RedshiftWriter.unloadData(RedshiftWriter.scala:330)
	at io.github.spark_redshift_community.spark.redshift.RedshiftWriter.saveToRedshift(RedshiftWriter.scala:405)
	at io.github.spark_redshift_community.spark.redshift.DefaultSource.createRelation(DefaultSource.scala:110)
	at org.apache.spark.sql.execution.datasources.SaveIntoDataSourceCommand.run(SaveIntoDataSourceCommand.scala:45)
	at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult$lzycompute(commands.scala:75)
	at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult(commands.scala:73)
	at org.apache.spark.sql.execution.command.ExecutedCommandExec.executeCollect(commands.scala:84)
	at org.apache.spark.sql.execution.QueryExecution$$anonfun$eagerlyExecuteCommands$1.$anonfun$applyOrElse$1(QueryExecution.scala:110)
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$5(SQLExecution.scala:103)
	at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:163)
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$1(SQLExecution.scala:90)
	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:775)
	at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:64)
	at org.apache.spark.sql.execution.QueryExecution$$anonfun$eagerlyExecuteCommands$1.applyOrElse(QueryExecution.scala:110)
	at org.apache.spark.sql.execution.QueryExecution$$anonfun$eagerlyExecuteCommands$1.applyOrElse(QueryExecution.scala:106)
	at org.apache.spark.sql.catalyst.trees.TreeNode.$anonfun$transformDownWithPruning$1(TreeNode.scala:481)
	at org.apache.spark.sql.catalyst.trees.CurrentOrigin$.withOrigin(TreeNode.scala:82)
	at org.apache.spark.sql.catalyst.trees.TreeNode.transformDownWithPruning(TreeNode.scala:481)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.org$apache$spark$sql$catalyst$plans$logical$AnalysisHelper$$super$transformDownWithPruning(LogicalPlan.scala:30)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.transformDownWithPruning(AnalysisHelper.scala:267)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.transformDownWithPruning$(AnalysisHelper.scala:263)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.transformDownWithPruning(LogicalPlan.scala:30)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.transformDownWithPruning(LogicalPlan.scala:30)
	at org.apache.spark.sql.catalyst.trees.TreeNode.transformDown(TreeNode.scala:457)
	at org.apache.spark.sql.execution.QueryExecution.eagerlyExecuteCommands(QueryExecution.scala:106)
	at org.apache.spark.sql.execution.QueryExecution.commandExecuted$lzycompute(QueryExecution.scala:93)
	at org.apache.spark.sql.execution.QueryExecution.commandExecuted(QueryExecution.scala:91)
	at org.apache.spark.sql.execution.QueryExecution.assertCommandExecuted(QueryExecution.scala:128)
	at org.apache.spark.sql.DataFrameWriter.runCommand(DataFrameWriter.scala:848)
	at org.apache.spark.sql.DataFrameWriter.saveToV1Source(DataFrameWriter.scala:382)
	at org.apache.spark.sql.DataFrameWriter.saveInternal(DataFrameWriter.scala:355)
	at org.apache.spark.sql.DataFrameWriter.save(DataFrameWriter.scala:247)
```

Looking at the code it seems that the problem is related to these lines:

https://github.com/spark-redshift-community/spark-redshift/blob/1f12470735c17022ae2dd68d0963649577928dde/src/main/scala/io/github/spark_redshift_community/spark/redshift/RedshiftWriter.scala#L313-L332

At L313 `fs` is defined on `s3a` protocol instead whereas `manifestPath` at L329 is defined on a sanitized version that relies on `s3` protocol causing my error

Is that correct?

### PR DESCRIPTION
**Pull request by conker84**
This fixes #103

### CODE REVIEW

**Comment by conker84** [comment]
@jsleight @lucagiovagnoli @88manpreet hi everybody, can someone review this?

File: src/main/scala/io/github/spark_redshift_community/spark/redshift/RedshiftWriter.scala [hunk 1]
```diff
@@ -326,7 +326,7 @@ private[redshift] class RedshiftWriter(
         s"""{"url":"$sanitizedTempDir/$file", "mandatory":true}"""
       }
       val manifest = s"""{"entries": [${manifestEntries.mkString(",\n")}]}"""
-      val manifestPath = sanitizedTempDir + "/manifest.json"
+      val manifestPath = tempDir + "manifest.json"
```
Comment (lucagiovagnoli) [inline src/main/scala/io/github/spark_redshift_community/spark/redshift/RedshiftWriter.scala:329]:
```
      // It's possible that tempDir contains AWS access keys. We shouldn't save those credentials to
// S3, so let's first sanitize `tempdir` and make sure that it uses the s3:// scheme:
```

You should still use a sanitized version of the url here. Also it'd  be useful to the reviewers if you can print out the detailed outcome of your tests (manual and automatic)

Please ask @88manpreet or someone else to review, as I am not a contributor to this anymore.

Comment (conker84) [reply src/main/scala/io/github/spark_redshift_community/spark/redshift/RedshiftWriter.scala:329]: I think that your statement is not correct, the `sanitizedTempDir` is necessary for the `COPY` command and that part is still as before (see L326).

@88manpreet can you please review it? Thank you so much!

Comment (jsleight) [reply src/main/scala/io/github/spark_redshift_community/spark/redshift/RedshiftWriter.scala:329]: The manifestPath is the return value of `unloadData` and eventually feeds into `RedshiftWriter.copySql` via `RedshiftWriter.doRedshiftLoad`, which itself does `val fixedUrl = Utils.fixS3Url(manifestUrl)`.  So I think this change is a no-op except for it not stripping out the credentials now?  

And thematically, the manifestPath gets passed into the Redshift Copy command, redshift would want the s3:// format.

Have you tested that this change solves your original problem?

Comment (jsleight) [reply src/main/scala/io/github/spark_redshift_community/spark/redshift/RedshiftWriter.scala:329]: Perhaps the change we need is that `val fsDataOut = fs.create(new Path(manifestPath))` needs to be `val fsDataOut = fs.create(new Path(tempDir))`, though we probably still want to strip the credentials out of tempDir first.  I think we would want to leave the return value unchanged though?

**Comment by conker84** [comment]
@jsleight @88manpreet ping

File: src/main/scala/io/github/spark_redshift_community/spark/redshift/RedshiftWriter.scala (hunk 1 above)
Comment (conker84) [reply src/main/scala/io/github/spark_redshift_community/spark/redshift/RedshiftWriter.scala:329]: @jsleight I tested the changed and it do fixes the problem That's why I proposed it

Comment (conker84) [reply src/main/scala/io/github/spark_redshift_community/spark/redshift/RedshiftWriter.scala:329]: I need to test if it works.
I'll be back ASAP

**Comment by conker84** [comment]
@jsleight @88manpreet I updated the PR

File: src/main/scala/io/github/spark_redshift_community/spark/redshift/RedshiftWriter.scala (hunk 1 above)
Comment (conker84) [reply src/main/scala/io/github/spark_redshift_community/spark/redshift/RedshiftWriter.scala:329]: @jsleight I'm quite sure that the comment refers to **this** line:
https://github.com/spark-redshift-community/spark-redshift/pull/104/files#diff-2880007d44472ced415970c8cb591813fd8f544ff81fade418334893c4f86355R326

And we don't save the credentials to S3, so I guess that we're good.
Wdyt?

**Comment by aatallah2** [comment]
We have the exact same issue, and made a similar fix locally. I'm not a reviewer, but I can attest that this is an issue with the latest AWS library dependencies.

The problem is that the manifest path must be in s3a, s3n,... format, whereas the path *inside* the manifest must be in s3:// format.

**Comment by conker84** [comment]
@jsleight please lemme know if you want to fix this

**Comment by jsleight** [comment]
I just merged #110 which is a duplicate of this issue.

**Pull request closed without merging**